javalib/src/main/scala/java/net/InetAddress.scala

```diff
@@ -592,105 +204,562 @@ object InetAddress extends InetAddressBase {
+            new IOException(gaiMsg)
+        throw ex
+      } else
+        try {
+          val preferIPv6 = SocketHelpers.getPreferIPv6Addresses()
+          findPreferrredAddrinfo(preferIPv6, !addrinfo) match {
+            case None =>
+              throw new UnknownHostException(s"${host}: Name does not resolve")
+            case Some(ai) => InetAddress(ai, host, isNumeric = false)
+          }
+        } finally {
+          freeaddrinfo(!addrinfo)
+        }
   }
 
-  def getAddress() = ipAddress.clone
+  /* Fully Qualified Domain Name which may or may not be the same as the
+   * canonical name.
+   */
+  private def getFullyQualifiedDomainName(
+      ipByteArray: Array[Byte]
+  ): Option[String] = {
+    /* MAXDNAME is the largest size of a Fully Qualified Domain Name.
+     * It is defined in:
+     *   https://github.com/openbsd/src/blob/master/include/arpa/nameser.h
+     *
+     * That URL says: "Define constants based on rfc883".
+     * These are direct name (bind) server definitions.
+     *
+     * This is larger than the length of individual segments because there
+     * can be multiple segments of 256. Two56.Two56.Two56.com
+     *
+     * On many BSD derived systems, this value is defined as (non-POSIX)
+     * NI_MAXHOST.
+     *   https://man7.org/linux/man-pages/man3/getnameinfo.3.html
+     *
+     * RFC 2181, section "Name syntax" states:
+     *   The length of any one label is limited to between 1 and 63 octets.
+     *   A full domain name is limited to 255 octets (including the
+     *   separators).
+     *
+     * A CString needs one more space for its terminal NUL.
+     *
+     * Use the larger MAXDNAME here, the extra space is not _all_ that
+     * expensive, and it is not used for long.
+     */
 
-  override def equals(obj: Any): Boolean = {
-    if (obj == null || obj.getClass != this.getClass) {
-      false
-    } else {
-      val objIPAddress = obj.asInstanceOf[InetAddress].getAddress()
-      objIPAddress.indices.forall(i => objIPAddress(i) == ipAddress(i))
+    val MAXDNAME = 1025.toUInt /* maximum presentation domain name */
+
+    def tailorSockaddr(ipBA: Array[Byte], addr: Ptr[sockaddr]): Unit = {
+      val from =
+        ipBA.asInstanceOf[scala.scalanative.runtime.Array[Byte]].at(0)
+
+      // By contract the 'sockaddr' argument passed in is cleared/all_zeros.
+      if (ipBA.length == 16) {
+        val v6addr = addr.asInstanceOf[Ptr[sockaddr_in6]]
+        v6addr.sin6_family = AF_INET6.toUShort
+        // because the FQDN scope is Global, no need to set sin6_scope_id
+        val dst = v6addr.sin6_addr.at1.at(0).asInstanceOf[Ptr[Byte]]
+        memcpy(dst, from, 16.toUInt)
+      } else if (ipBA.length == 4) {
+        val v4addr = addr.asInstanceOf[Ptr[sockaddr_in]]
+        v4addr.sin_family = AF_INET.toUShort
+        v4addr.sin_addr = !(from.asInstanceOf[Ptr[in_addr]]) // Structure copy
+      } else {
+        throw new IOException("Invalid ipAddress length: ${ipBa.length}")
```

Comment:
```suggestion
        throw new IOException(s"Invalid ipAddress length: ${ipBa.length}")
```

Reply:
Arman, thank you for the catch!  I will implement your suggestion.